### ISSUE TITLE
`stack_snapshot` get invalidated too often

### ISSUE THREAD

**Issue by regnat** [tweag/rules_haskell#1096]
**Describe the bug**

Dependencies built with `stack_packages` or `haskell_cabal_package` gets
invalidated at least each time the `PATH` changes due to the use of
`use_default_shell_env`.

**To Reproduce**

From the `examples` directory:

```sh
# Build a random cabal package
nix-shell --pure --run "bazel build @stackage//:random"
# Rebuild it again and see that it's cached
nix-shell --pure --run "bazel build @stackage//:random"
# Add `hello` to the `buildInputs` list in `shell.nix`
# Rebuild the random package and see that it isn't cached anymore
nix-shell --pure --run "bazel build @stackage//:random"
```

**Expected behavior**

A change in the environment shouldn't trigger a cache miss.

**Environment**
 - OS name + version: nixos unstable
 - Bazel version: 0.28.0 (from the example's shell.nix)
 - Version of the rules: latest master (79e2c5316bef9873f6d6a155f1f14f3d3ae39967)

**Comment by mboes** [comment]
Did you mean `stack_snapshot`?

**Comment by regnat** [comment]
> Did you mean `stack_snapshot`?

Oh yes, sorry, I'll update the title

**Comment by mboes** [comment]
> each time the PATH changes due to the use of use_default_shell_env.

This is a side-effect of not having a shell toolchain. Cabal assumes quite a bit from the environment (in terms of available commands).

**Comment by jkachmar** [comment]
Should #1097 and #1117 (which have since been merged) automatically address the problem, and should this issue be closed accordingly?

Or is there a need for some sort of regression test to make sure `stack_snapshot` dependencies are always cached?

**Comment by aherrmann** [comment]
This should be fixed now. I've verified that the following does not cause a rebuild, while it did before those PRs were merged.
```
$ nix-shell --pure --run 'bazel build @stackage//:void'
$ nix-shell --pure --run 'PATH=foo:$PATH bazel build @stackage//:void'
```
